src/Features/CSharp/Portable/CodeRefactorings/InlineMethod/CSharpInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,194 @@
+                   || isNullConditionalInvocationExpression
+                   || expressionNode.IsKind(SyntaxKind.ObjectCreationExpression)
+                   || expressionNode is AssignmentExpressionSyntax
+                   || expressionNode.IsKind(SyntaxKind.PreIncrementExpression)
+                   || expressionNode.IsKind(SyntaxKind.PreDecrementExpression)
+                   || expressionNode.IsKind(SyntaxKind.PostIncrementExpression)
+                   || expressionNode.IsKind(SyntaxKind.PostDecrementExpression)
+                   || expressionNode.IsKind(SyntaxKind.AwaitExpression);
+        }
+
+        protected override bool TryGetInlineSyntaxNodeAndReplacementNodeForDelegate(
+            InvocationExpressionSyntax calleeInvocationNode,
+            IMethodSymbol calleeMethodSymbol,
+            ExpressionSyntax inlineExpressionNode,
+            StatementSyntax statementContainsCallee,
+            SyntaxGenerator syntaxGenerator,
+            out SyntaxNode? inlineSyntaxNode,
+            out SyntaxNode? syntaxNodeToReplace)
+        {
+            inlineSyntaxNode = null;
+            syntaxNodeToReplace = null;
+            if (statementContainsCallee is LocalDeclarationStatementSyntax localDeclarationSyntax
+                && IsVariableInitializerInLocalDeclarationSyntax(calleeInvocationNode, localDeclarationSyntax)
+                && localDeclarationSyntax.Declaration.Type.IsVar)
+            {
+                // Example:
+                // Before:
+                // void Caller() { var x = Callee(); }
+                // Action Callee() { return () => {}; }
+                //
+                // After inline it should be
+                // void Caller() { Action x = () => {};}
+                // Action Callee() { return () => {}; }
+                // 'var' can't be used for delegate
+                inlineSyntaxNode = UseExplicitTypeAndReplaceInitializerForDeclarationSyntax(
+                    localDeclarationSyntax,
+                    syntaxGenerator,
+                    calleeMethodSymbol.ReturnType,
+                    calleeInvocationNode,
+                    inlineExpressionNode);
+
+                syntaxNodeToReplace = statementContainsCallee;
+                return true;
+            }
+
+            // Example:
+            // Before:
+            // void Caller() { var x = Callee()(); }
+            // Func<int> Callee() { return () => 1; }
+            // After:
+            // void Caller() { var x = ((Func<int>)(() => 1))(); }
+            // Func<int> Callee() { return () => 1; }
+            // Cast expression is needed for lambda
+            if (calleeInvocationNode.Parent?.IsKind(SyntaxKind.InvocationExpression) == true)
+            {
+                inlineSyntaxNode = SyntaxFactory.CastExpression(
+                    calleeMethodSymbol.ReturnType.GenerateTypeSyntax(allowVar: false),
+                    inlineExpressionNode.Parenthesize()).Parenthesize();
```

Comment:
oh wait.  you are casting+parenthesizing.  so i don't see why you need the above check at all.

Reply:
I used to think it will only apply to invocation expression. But sounds like the simplifier can do a lot work here to make sure the cast is necessary.